### ISSUE TITLE
How to create a container with the same root fs as the host?

### ISSUE THREAD

**Issue by planetA** [opencontainers/runc#1909]
Hello,

I'm trying to create a container that runs in its own namespaces (mount, net, uts, etc.). But the special thing I want to achieve is that it has the same view on the filesystem as the root, unless it tries to modify its own VFS.

For that I try to specify path to rootfs as "/", but the creation of the container does not work this way and fails with "invalid argument":

> sudo ./test-runc hostname
> 2018/10/13 10:34:34 exec: "hostname": executable file not found in $PATH
> 2018/10/13 12:34:34 Failed to launch the container: container_linux.go:336: starting container process caused "exec: \"hostname\": executable file not found in $PATH"

Here is the full example I tried to run:  [test-runc](https://gist.github.com/planetA/64b032d186329c95721cb3ad867e12ab)

I still want to run the container in a separate mount namespace. For example, to be able to remount /proc, but I still want the rest of the host VFS be visible from the inside of the container.

**Comment by mikebrow** [comment]
instead of changing your rootfs path... add a /host mount pointing to root in your config.json. Cheers.

**Comment by planetA** [comment]
@mikebrow, thank you for the response, but could you be a bit more elaborate?

Do you mean creating a directory "/host" that is a bind mount to "/"? And then Rootfs parameter is "/host"?

**Comment by mikebrow** [comment]
No. **Instead** of changing your rootfs parameter/path... add a mount who's source is root and who's destination is /host or some other path that is not used by your container. Put another way, no I do not believe you can set the rootfs of the container to the host root, but you can mount the host's root as a non root directory in the container.

**Comment by sigmonsays** [comment]
I too would like this capability. Is there a way to achieve this?

I have setup a series of bind mounts on the host but it is not that flexible.

**Comment by mikebrow** [comment]
@planetA @sigmonsays any more help needed on this one? 

Summary, you can't set a container's rootfs to the host's root. But you can mount the host root as a non root directory in the container.

**Comment by planetA** [comment]
@mikebrow no

**Comment by qbing** [comment]
@planetA  Have you found the solution?

**Comment by planetA** [comment]
@qbing Not really

**Comment by Ragnarliu** [comment]
@planetA 
   I  also want to mount host "/" to  contianer "/" , it can work but when i disable the mount namespace there will have some issues. you can refer to the following link:
  https://groups.google.com/a/opencontainers.org/forum/#!searchin/dev/mount$20namespace%7Csort:date/dev/p10bq-kXODk/obkqBRdxCQAJ

**Comment by yozel** [comment]
If anyone still interested, I was able to achieve that by using `mount` namespace and `rbind` mount together. Also you need an empty initial rootfs, I've created a directory named `empty-directory `for that.

```json
{
	"ociVersion": "1.0.2-dev",
	"process": {
		"terminal": true,
		"user": {
			"uid": 1031682,
			"gid": 100
		},
		"args": [
			"sh"
		],
		"env": [
			"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
			"TERM=xterm"
		],
		"cwd": "/",
		"rlimits": [
			{
				"type": "RLIMIT_NOFILE",
				"hard": 1024,
				"soft": 1024
			}
		],
		"noNewPrivileges": true
	},
	"root": {
		"path": "empty-directory",
		"readonly": false
	},
	"hostname": "runc",
	"mounts": [
		{
			"destination": "/",
			"type": "rbind",
			"source": "/",
			"options": [
				"rw",
				"rbind"
			]
		}
	],
	"linux": {
		"resources": {
			"devices": [
				{
					"allow": true,
					"access": "rwm"
				}
			]
		},
		"namespaces": [
			{
				"type": "pid"
			},
			{
				"type": "network"
			},
			{
				"type": "ipc"
			},
			{
				"type": "uts"
			},
			{
				"type": "mount"
			}
		]
	}
}
```

**Comment by yozel** [comment]
If anyone still interested, I was able to achieve that by using `mount` namespace and `rbind` mount together. Also you need an empty initial rootfs, I've created a directory named `empty-directory `for that.

```json
{
	"ociVersion": "1.0.2-dev",
	"process": {
		"terminal": true,
		"user": {
			"uid": 1031682,
			"gid": 100
		},
		"args": [
			"sh"
		],
		"env": [
			"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
			"TERM=xterm"
		],
		"cwd": "/",
		"rlimits": [
			{
				"type": "RLIMIT_NOFILE",
				"hard": 1024,
				"soft": 1024
			}
		],
		"noNewPrivileges": true
	},
	"root": {
		"path": "empty-directory",
		"readonly": false
	},
	"hostname": "runc",
	"mounts": [
		{
			"destination": "/",
			"type": "rbind",
			"source": "/",
			"options": [
				"rw",
				"rbind"
			]
		}
	],
	"linux": {
		"resources": {
			"devices": [
				{
					"allow": true,
					"access": "rwm"
				}
			]
		},
		"namespaces": [
			{
				"type": "pid"
			},
			{
				"type": "network"
			},
			{
				"type": "ipc"
			},
			{
				"type": "uts"
			},
			{
				"type": "mount"
			}
		]
	}
}
```

**Comment by aeronwang** [comment]
[yozel](https://github.com/yozel) Thanks. It works but it has problems for example /dev/pts will have problem even after the container has removed. Generically some resources are not designed to be shared - as kernel does not isolate these yet. Or do you have any workaround? maybe bind mount most host dirs to container's rootfs but not these problematic ones?
